README.md
# Juicy Pixels Extra

[![License BSD3](https://img.shields.io/badge/license-BSD3-brightgreen.svg)](http://opensource.org/licenses/BSD-3-Clause)
[![Hackage](https://img.shields.io/hackage/v/JuicyPixels-extra.svg?style=flat)](https://hackage.haskell.org/package/JuicyPixels-extra)
[![Stackage Nightly](http://stackage.org/package/JuicyPixels-extra/badge/nightly)](http://stackage.org/nightly/package/JuicyPixels-extra)
[![Stackage LTS](http://stackage.org/package/JuicyPixels-extra/badge/lts)](http://stackage.org/lts/package/JuicyPixels-extra)
![CI](https://github.com/mrkkrp/JuicyPixels-extra/workflows/CI/badge.svg?branch=master)

A collection of functions to scale, crop, flip images with JuicyPixels.

## Contribution

Issues, bugs, and questions may be reported in [the GitHub issue tracker for
this project](https://github.com/mrkkrp/JuicyPixels-extra/issues).

Pull requests are also welcome.

## License

Copyright © 2016–present Mark Karpov

Distributed under BSD 3 clause license.


CHANGELOG.md
## Juicy Pixels Extra 0.6.0

* Added new functions `trim` and `square`. [See PR
  34](https://github.com/mrkkrp/JuicyPixels-extra/pull/34).

## Juicy Pixels Extra 0.5.2

* Replacing CPP preprocessing rules with expicit SPECIALIZE pragmas. [See PR
  26](https://github.com/mrkkrp/JuicyPixels-extra/pull/26).

## Juicy Pixels Extra 0.5.1

* `scaleBilinear` no longer hangs when the target width or height is 0.
  [Issue 24](https://github.com/mrkkrp/JuicyPixels-extra/issues/24).

## Juicy Pixels Extra 0.5.0

* Changed how edge pixels are calculated when scaling via bilinear
  interpolation is used. [See PR
  21](https://github.com/mrkkrp/JuicyPixels-extra/pull/21).

* Dropped support for GHC 8.2 and older.

## Juicy Pixels Extra 0.4.1

* Fixed left rotation so that `rotateLeft90 . rotateRight90` is now identity
  function.

* Dropped support for GHC 7.10.

## Juicy Pixels Extra 0.4.0

* Fix `scaleBilinear` so it works correctly with images that have 16-bit
  components.

## Juicy Pixels Extra 0.3.0

* Made the function `scaleBilinear` polymorphic in pixel type.

## Juicy Pixels Extra 0.2.2

* Fixed a bug in the `crop` function related to incorrect calculation of new
  height in some cases.

## Juicy Pixels Extra 0.2.1

* Documentation improvements.

## Juicy Pixels Extra 0.2.0

* Added `rotate180`, `beside`, and `below`.

## Juicy Pixels Extra 0.1.1

* Added benchmarks.

* Considerably improved performance of all functions (they are about ×15
  faster now).

## Juicy Pixels Extra 0.1.0

* Initial release.


tests/Codec/Picture/ExtraSpec.hs
{-# LANGUAGE FlexibleContexts #-}

module Codec.Picture.ExtraSpec
  ( main,
    spec,
  )
where

import Codec.Picture
import Codec.Picture.Extra
import Control.Monad
import Data.Function (on)
import Foreign (Storable)
import Test.Hspec

main :: IO ()
main = hspec spec

spec :: Spec
spec = do
  describe "scaleBilinear" scaleBilinearSpec
  describe "crop" cropSpec
  describe "trim" trimSpec
  describe "flipHorizontally" flipHorizontallySpec
  describe "flipVertically" flipVerticallySpec
  describe "rotateLeft90Spec" rotateLeft90Spec
  describe "rotateRight90Spec" rotateRight90Spec
  describe "rotate180" rotate180Spec
  describe "beside" besideSpec
  describe "below" belowSpec
  describe "square" squareSpec

scaleBilinearSpec :: Spec
scaleBilinearSpec = do
  context "when called with orginial dimensions" $
    it "produces the same image" $
      checkWithFiles
        (scaleBilinear 512 512)
        "data-examples/macaque.png"
        "data-examples/macaque.png"
  context "when we scale down" $
    it "produces correct image" $
      checkWithFiles
        (scaleBilinear 100 100)
        "data-examples/macaque.png"
        "data-examples/macaque-scaled-down.png"
  context "when we scale up" $
    it "produces correct image" $
      checkWithFiles
        (scaleBilinear 600 600)
        "data-examples/macaque.png"
        "data-examples/macaque-scaled-up.png"
  context "when we scale to 1 width" $
    it "produces correct image" $
      checkWithFiles
        (scaleBilinear 1 512)
        "data-examples/macaque.png"
        "data-examples/macaque-width-1.png"
  context "when we scale to 1 height" $
    it "produces correct image" $
      checkWithFiles
        (scaleBilinear 512 1)
        "data-examples/macaque.png"
        "data-examples/macaque-height-1.png"
  context "when we scale to 0x0" $
    it "produces the empty image" $
      checkWithFiles
        (scaleBilinear 0 0)
        "data-examples/macaque.png"
        "data-examples/empty.png"

cropSpec :: Spec
cropSpec = do
  context "when we pass arguments within image size (square)" $
    it "produces correct image" $
      checkWithFiles
        (crop 211 210 178 191)
        "data-examples/macaque.png"
        "data-examples/macaque-cropped.png"
  context "when we pass arguments within image size (vertical)" $
    it "produces correct image" $
      checkWithFiles
        (crop 0 512 512 512)
        "data-examples/macaque-below.png"
        "data-examples/macaque.png"

trimSpec :: Spec
trimSpec = do
  context "when we pass an image without transparency" $
    it "does nothing" $
      checkWithFiles
        trim
        "data-examples/macaque.png"
        "data-examples/macaque.png"
  context "when passing an image with transparency" $ do
    it "does nothing if there are no fully transparent edges" $
      checkWithFilesAlpha
        trim
        "data-examples/macaque-transparent-cropped.png"
        "data-examples/macaque-transparent-cropped.png"
    it "removed the transparent edges" $
      checkWithFilesAlpha
        trim
        "data-examples/macaque-transparent.png"
        "data-examples/macaque-transparent-cropped.png"
  context "when passing a fully transparent image" $
    it "returns an empty image" $ do
      (Right (ImageRGBA8 original)) <- readImage "data-examples/fully-transparent.png"
      let trimmed = trim original
      imageWidth trimmed `shouldBe` 1
      imageHeight trimmed `shouldBe` 1

flipHorizontallySpec :: Spec
flipHorizontallySpec =
  context "when we flip horizontally" $
    it "produces correct image" $
      checkWithFiles
        flipHorizontally
        "data-examples/macaque.png"
        "data-examples/macaque-horizontal-flip.png"

flipVerticallySpec :: Spec
flipVerticallySpec =
  context "when we flip vertically" $
    it "produces correct image" $
      checkWithFiles
        flipVertically
        "data-examples/macaque.png"
        "data-examples/macaque-vertical-flip.png"

rotateLeft90Spec :: Spec
rotateLeft90Spec =
  context "when we rotate to the left by 90°" $
    it "produces correct image" $
      checkWithFiles
        rotateLeft90
        "data-examples/macaque.png"
        "data-examples/macaque-left-rotated.png"

rotateRight90Spec :: Spec
rotateRight90Spec =
  context "when we rotate to the right by 90°" $
    it "produces correct image" $
      checkWithFiles
        rotateRight90
        "data-examples/macaque.png"
        "data-examples/macaque-right-rotated.png"

rotate180Spec :: Spec
rotate180Spec =
  context "when we rotate by 180°" $
    it "produces correct image" $
      checkWithFiles
        rotate180
        "data-examples/macaque.png"
        "data-examples/macaque-180-rotated.png"

besideSpec :: Spec
besideSpec =
  context "when we place images beside each other" $
    it "produces correct image" $
      checkWithFiles
        (\x -> beside [x, x])
        "data-examples/macaque.png"
        "data-examples/macaque-beside.png"

belowSpec :: Spec
belowSpec =
  context "when we place images below each other" $
    it "produces correct image" $
      checkWithFiles
        (\x -> below [x, x])
        "data-examples/macaque.png"
        "data-examples/macaque-below.png"

squareSpec :: Spec
squareSpec = do
  context "when we pass an already square image" $ do
    it "does nothing" $
      checkWithFiles
        (square $ PixelRGB8 0 0 0)
        "data-examples/macaque.png"
        "data-examples/macaque.png"
    it "does nothing (alpha)" $
      checkWithFilesAlpha
        (square $ PixelRGBA8 0 0 0 0)
        "data-examples/macaque-transparent.png"
        "data-examples/macaque-transparent.png"
  context "when we pass a non-square image" $ do
    it "adds the required filler to wider than heigh image" $ do
      (Right (ImageRGB8 original)) <- readImage "data-examples/macaque-beside.png"
      let squared = square (PixelRGB8 0 0 0) original
      imageWidth squared `shouldBe` 1024
      imageHeight squared `shouldBe` 1024
    it "adds the required filler to higher than wide image" $ do
      (Right (ImageRGB8 original)) <- readImage "data-examples/macaque-below.png"
      let squared = square (PixelRGB8 0 0 0) original
      imageWidth squared `shouldBe` 1024
      imageHeight squared `shouldBe` 1024

-- | Run a transforming on the image loaded from a file and compare the
-- resulting image with the contents of another file.
checkWithFiles ::
  -- | Transformation to test
  (Image PixelRGB8 -> Image PixelRGB8) ->
  -- | Where to get the original image
  FilePath ->
  -- | Where to get the image to compare with
  FilePath ->
  Expectation
checkWithFiles f opath fpath = do
  (Right (ImageRGB8 original)) <- readImage opath
  (Right (ImageRGB8 result)) <- readImage fpath
  f original `blindlySatisfy` sameImage result

-- | `checkWithFiles` for images with an alpha channel.
checkWithFilesAlpha ::
  -- | Transformation to test
  (Image PixelRGBA8 -> Image PixelRGBA8) ->
  -- | Where to get the original image
  FilePath ->
  -- | Where to get the image to compare with
  FilePath ->
  Expectation
checkWithFilesAlpha f opath fpath = do
  (Right (ImageRGBA8 original)) <- readImage opath
  (Right (ImageRGBA8 result)) <- readImage fpath
  f original `blindlySatisfy` sameImage result

-- | The same as 'shouldSatisfy', but doesn't care if its argument is an
-- instance of 'Show' or not.
blindlySatisfy :: a -> (a -> Bool) -> Expectation
v `blindlySatisfy` p =
  unless (p v) (expectationFailure "predicate failed")

-- | The equality test for images.
sameImage :: (Eq a, Eq (PixelBaseComponent a), Storable (PixelBaseComponent a)) => Image a -> Image a -> Bool
sameImage a b =
  ((==) `on` imageWidth) a b
    && ((==) `on` imageHeight) a b
    && ((==) `on` imageData) a b


Codec/Picture/Extra.hs
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE ScopedTypeVariables #-}

-- |
-- Module      :  Codec.Picture.Extra
-- Copyright   :  © 2016–present Mark Karpov
-- License     :  BSD 3 clause
--
-- Maintainer  :  Mark Karpov <markkarpov92@gmail.com>
-- Stability   :  experimental
-- Portability :  portable
--
-- A collection of functions to scale, crop, flip images with JuicyPixels.
module Codec.Picture.Extra
  ( -- * Scaling
    scaleBilinear,

    -- * Cropping
    crop,
    trim,

    -- * Rotation
    flipHorizontally,
    flipVertically,
    rotateLeft90,
    rotateRight90,
    rotate180,

    -- * Other
    beside,
    below,
    square,
  )
where

import Codec.Picture
import qualified Codec.Picture.Types as M
import Control.Monad.ST
import Data.List (find, foldl1')
import Data.Maybe (fromMaybe)

-- | Scale an image using bi-linear interpolation.
scaleBilinear ::
  ( Pixel a,
    Bounded (PixelBaseComponent a),
    Integral (PixelBaseComponent a)
  ) =>
  -- | Desired width
  Int ->
  -- | Desired height
  Int ->
  -- | Original image
  Image a ->
  -- | Scaled image
  Image a
scaleBilinear width height img@Image {..}
  | width <= 0 || height <= 0 =
      generateImage (error "scaleBilinear: absurd") (max 0 width) (max 0 height)
  | otherwise = runST $ do
      mimg <- M.newMutableImage width height
      let sx, sy :: Float
          sx = fromIntegral imageWidth / fromIntegral width
          sy = fromIntegral imageHeight / fromIntegral height
          go x' y'
            | x' >= width = go 0 (y' + 1)
            | y' >= height = M.unsafeFreezeImage mimg
            | otherwise = do
                let xf = fromIntegral x' * sx
                    yf = fromIntegral y' * sy
                    x, y :: Int
                    x = floor xf
                    y = floor yf
                    δx = xf - fromIntegral x
                    δy = yf - fromIntegral y
                    pixelAt' i j =
                      pixelAt img (min (pred imageWidth) i) (min (pred imageHeight) j)
                writePixel mimg x' y' $
                  mulp (pixelAt' x y) ((1 - δx) * (1 - δy))
                    `addp` mulp (pixelAt' (x + 1) y) (δx * (1 - δy))
                    `addp` mulp (pixelAt' x (y + 1)) ((1 - δx) * δy)
                    `addp` mulp (pixelAt' (x + 1) (y + 1)) (δx * δy)
                go (x' + 1) y'
      go 0 0
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.PixelRGBA16 -> Image M.PixelRGBA16 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.PixelRGBA8 -> Image M.PixelRGBA8 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.PixelCMYK16 -> Image M.PixelCMYK16 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.PixelCMYK8 -> Image M.PixelCMYK8 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.PixelYCbCr8 -> Image M.PixelYCbCr8 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.PixelRGB16 -> Image M.PixelRGB16 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.PixelYCbCrK8 -> Image M.PixelYCbCrK8 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.PixelRGB8 -> Image M.PixelRGB8 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.PixelYA16 -> Image M.PixelYA16 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.PixelYA8 -> Image M.PixelYA8 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.Pixel32 -> Image M.Pixel32 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.Pixel16 -> Image M.Pixel16 #-}
{-# SPECIALIZE scaleBilinear :: Int -> Int -> Image M.Pixel8 -> Image M.Pixel8 #-}

mulp :: (Pixel a, Integral (PixelBaseComponent a)) => a -> Float -> a
mulp pixel x = colorMap (floor . (* x) . fromIntegral) pixel
{-# INLINE mulp #-}

addp ::
  forall a.
  ( Pixel a,
    Bounded (PixelBaseComponent a),
    Integral (PixelBaseComponent a)
  ) =>
  a ->
  a ->
  a
addp = mixWith (const f)
  where
    f x y =
      fromIntegral $
        (maxBound :: PixelBaseComponent a) `min` (fromIntegral x + fromIntegral y)
{-# INLINE addp #-}

-- | Crop an image. If the supplied coordinates are greater than the size of
-- the image, the image boundaries are used instead.
crop ::
  Pixel a =>
  -- | Index (X axis) of first pixel to include
  Int ->
  -- | Index (Y axis) of first pixel to include
  Int ->
  -- | Width of resulting image
  Int ->
  -- | Height of resulting image
  Int ->
  -- | Original image
  Image a ->
  -- | Cropped image
  Image a
crop x' y' w' h' img@Image {..} =
  generateImage gen w h
  where
    gen i j = pixelAt img (x + i) (y + j)
    x = min (imageWidth - 1) x'
    y = min (imageHeight - 1) y'
    w = min (imageWidth - x) w'
    h = min (imageHeight - y) h'
{-# INLINEABLE crop #-}

-- | Trim the completely transparent edges of an image.
--
-- @since 0.6.0
trim :: (Pixel a, Eq (PixelBaseComponent a)) => Image a -> Image a
trim img@Image {..} = crop left top width height img
  where
    isInvisible p = pixelOpacity p == 0
    isInvisibleRow y = all isInvisible $ flip (pixelAt img) y <$> [0 .. imageWidth - 1]
    isInvisibleCol x = all isInvisible $ pixelAt img x <$> [0 .. imageHeight - 1]

    top = fromMaybe 0 (find (not . isInvisibleRow) [0 .. imageHeight - 1])
    bottom = fromMaybe 0 (find (not . isInvisibleRow) [imageHeight - 1, imageHeight - 2 .. 0]) + 1
    height = bottom - top

    left = fromMaybe 0 (find (not . isInvisibleCol) [0 .. imageWidth - 1])
    right = fromMaybe 0 (find (not . isInvisibleCol) [imageWidth - 1, imageWidth - 2 .. 1]) + 1
    width = right - left
{-# INLINEABLE trim #-}

-- | Flip an image horizontally.
flipHorizontally :: Pixel a => Image a -> Image a
flipHorizontally img@Image {..} =
  generateImage gen imageWidth imageHeight
  where
    gen x = pixelAt img (imageWidth - 1 - x)
{-# INLINEABLE flipHorizontally #-}

-- | Flip an image vertically.
flipVertically :: Pixel a => Image a -> Image a
flipVertically img@Image {..} =
  generateImage gen imageWidth imageHeight
  where
    gen x y = pixelAt img x (imageHeight - 1 - y)
{-# INLINEABLE flipVertically #-}

-- | Rotate an image to the left by 90°.
rotateLeft90 :: Pixel a => Image a -> Image a
rotateLeft90 img@Image {..} =
  generateImage gen imageHeight imageWidth
  where
    gen x y = pixelAt img (imageWidth - 1 - y) x
{-# INLINEABLE rotateLeft90 #-}

-- | Rotate an image to the right by 90°.
rotateRight90 :: Pixel a => Image a -> Image a
rotateRight90 img@Image {..} =
  generateImage gen imageHeight imageWidth
  where
    gen x y = pixelAt img y (imageHeight - 1 - x)
{-# INLINEABLE rotateRight90 #-}

-- | Rotate an image by 180°, i.e flip both vertically and horizontally.
--
-- @since 0.2.0
rotate180 :: Pixel a => Image a -> Image a
rotate180 img@(Image w h _) = generateImage g w h
  where
    g x y = pixelAt img (w - 1 - x) (h - 1 - y)
{-# INLINEABLE rotate180 #-}

-- | Create an image by placing several images side by side. If the images
-- are of differnet heights the smallest height is used.
--
-- @since 0.2.0
beside :: Pixel a => [Image a] -> Image a
beside = foldl1' go
  where
    go :: Pixel a => Image a -> Image a -> Image a
    go img1@(Image w1 h1 _) img2@(Image w2 h2 _) =
      generateImage g (w1 + w2) h
      where
        g x
          | x < w1 = pixelAt img1 x
          | otherwise = pixelAt img2 (x - w1)
        h = min h1 h2
{-# INLINEABLE beside #-}

-- | Create an image by placing several images in a vertical stack. If the
-- images are of differnet widths the smallest width is used.
--
-- @since 0.2.0
below :: Pixel a => [Image a] -> Image a
below = foldl1' go
  where
    go :: Pixel a => Image a -> Image a -> Image a
    go img1@(Image w1 h1 _) img2@(Image w2 h2 _) =
      generateImage g w (h1 + h2)
      where
        g x y
          | y < h1 = pixelAt img1 x y
          | otherwise = pixelAt img2 x (y - h1)
        w = min w1 w2
{-# INLINEABLE below #-}

-- | Make an image a perfect square by adding filler around it.
--
-- @since 0.6.0
square :: Pixel a => a -> Image a -> Image a
square filler img@Image {..} =
  if imageWidth == imageHeight
    then img
    else generateImage gen size size
  where
    size = max imageWidth imageHeight
    extraWidth = size - imageWidth
    extraHeight = size - imageHeight
    offsetX = extraWidth `div` 2
    offsetY = extraHeight `div` 2
    gen i _ | i < offsetX = filler
    gen i _ | i >= imageWidth + offsetX = filler
    gen _ j | j < offsetY = filler
    gen _ j | j >= imageHeight + offsetY = filler
    gen i j = pixelAt img (i - offsetX) (j - offsetY)
{-# INLINEABLE square #-}


bench/Main.hs
module Main (main) where

import Codec.Picture
import Codec.Picture.Extra
import Criterion.Main

main :: IO ()
main =
  defaultMain
    [ bgroup
        "scaleBilinear"
        [ baction "512 × 512" (scaleBilinear 512 512) "data-examples/macaque.png",
          baction "256 × 256" (scaleBilinear 256 256) "data-examples/macaque.png",
          baction "1024 × 1024" (scaleBilinear 1024 1024) "data-examples/macaque.png"
        ],
      bgroup
        "crop"
        [ baction "on 256 horizontally" (crop 256 0 256 512) "data-examples/macaque.png",
          baction "on 265 vertically" (crop 0 256 512 256) "data-examples/macaque.png"
        ],
      bgroup
        "flipHorizontally"
        [ baction "512 × 512" flipHorizontally "data-examples/macaque.png",
          baction "100 × 100" flipHorizontally "data-examples/macaque-scaled-down.png"
        ],
      bgroup
        "flipVertically"
        [ baction "512 × 512" flipVertically "data-examples/macaque.png",
          baction "100 × 100" flipVertically "data-examples/macaque-scaled-down.png"
        ],
      bgroup
        "rotateLeft90"
        [ baction "512 × 512" rotateLeft90 "data-examples/macaque.png",
          baction "100 × 100" rotateLeft90 "data-examples/macaque-scaled-down.png"
        ],
      bgroup
        "rotateRight90"
        [ baction "512 × 512" rotateRight90 "data-examples/macaque.png",
          baction "100 × 100" rotateRight90 "data-examples/macaque-scaled-down.png"
        ],
      bgroup
        "rotate180"
        [ baction "512 × 512" rotate180 "data-examples/macaque.png",
          baction "100 × 100" rotate180 "data-examples/macaque-scaled-down.png"
        ],
      bgroup
        "beside"
        [ baction "1024 × 512" beside' "data-examples/macaque.png",
          baction "200 × 100" beside' "data-examples/macaque-scaled-down.png"
        ],
      bgroup
        "below"
        [ baction "512 × 1024" below' "data-examples/macaque.png",
          baction "100 × 200" below' "data-examples/macaque-scaled-down.png"
        ]
    ]
  where
    beside' img = beside [img, img]
    below' img = below [img, img]

-- | Run a benchmark given a function to benchmark and where to get image to
-- pass as the input.
baction ::
  -- | Name of the benchmark
  String ->
  -- | The transformation to perform
  (Image PixelRGB8 -> Image PixelRGB8) ->
  -- | Where to get the image to start with
  FilePath ->
  Benchmark
baction name f path = env getImg (bench name . nf f)
  where
    getImg = do
      (Right (ImageRGB8 img)) <- readImage path
      return img
